overlord/state/taskrunner.go

```diff
@@ -183,6 +200,30 @@ func (r *TaskRunner) AddBlocked(pred func(t *Task, running []*Task) bool) {
 	r.blocked = append(r.blocked, pred)
 }
 
+// AddHook registers a hook function for the given hook event type.
+func (r *TaskRunner) AddHook(hook HookFunc, evt HookEvent) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.hooks[evt] = append(r.hooks[evt], hook)
+}
+
+func (r *TaskRunner) checkExhaustion() {
+	for _, chg := range r.state.changes {
+		if chg.IsReady() || chg.Status() == WaitStatus {
+			// have we already reported exhaustion for that change?
+			if r.exhausted[chg.ID()] {
```

Comment:
if we check just one change per run(Task) do we need the cache?

Reply:
Nope, I could remove the cache, makes it so much nicer!